src/neo4j_genai/retrievers/text_to_cypher.py

```diff
@@ -0,0 +1,104 @@
+from neo4j_genai.types import (
+    LLMModel,
+    Neo4jDriverModel,
+    Neo4jSchemaModel,
+    TextToCypherRetrieverModel,
+    TextToCypherSearchModel,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class TextToCypherRetriever(Retriever):
+    """
+    Allows for the retrieval of records from a Neo4j database using natural language.
+    Converts a user's natural language query to a Cypher query using an LLM,
+    then retrieves records from a Neo4j database using the generated Cypher query
+    """
+    def __init__(
+        self, driver: neo4j.Driver, llm: LLM, neo4j_schema: Optional[str] = None
+    ) -> None:
+        try:
+            driver_model = Neo4jDriverModel(driver=driver)
+            llm_model = LLMModel(llm=llm)
+            neo4j_schema_model = (
+                Neo4jSchemaModel(neo4j_schema=neo4j_schema) if neo4j_schema else None
+            )
+            validated_data = TextToCypherRetrieverModel(
+                driver_model=driver_model,
+                llm_model=llm_model,
+                neo4j_schema_model=neo4j_schema_model,
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        super().__init__(validated_data.driver_model.driver)
+        self.llm = validated_data.llm_model.llm
+        self.neo4j_schema = (
+            validated_data.neo4j_schema_model.neo4j_schema
+            if validated_data.neo4j_schema_model
+            else get_schema(validated_data.driver_model.driver)
+        )
+
+    def search(
+        self, query_text: str, examples: Optional[list[str]] = None
+    ) -> list[neo4j.Record]:
+        """Converts query_text to a Cypher query using an LLM.
+           Retrieve records from a Neo4j database using the generated Cypher query.
+
+        Args:
+            query_text (str): The natural language query used to search the Neo4j database.
+            examples (Optional[list[str], optional): Optional user input/query pairs for the LLM to use as examples.
+
+        Raises:
+            ValidationError: If validation of the input arguments fail.
+
+        Returns:
+            list[neo4j.Record]: The results of the search query.
+        """
+        try:
+            validated_data = TextToCypherSearchModel(
+                query_text=query_text, examples=examples
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        prompt = TEXT_TO_CYPHER_PROMPT.format(
+            schema=self.neo4j_schema,
+            examples="\n".join(validated_data.examples)
+            if validated_data.examples
+            else "",
+            input=validated_data.query_text,
+        )
+        logger.debug("TextToCypherRetriever prompt: %s", prompt)
+
+        t2c_query = self.llm.invoke(prompt)
+        logger.debug("TextToCypherRetriever Cypher query: %s", t2c_query)
+
+        records, _, _ = self.driver.execute_query(query_=t2c_query)
```

Comment:
What would happen if the Cypher query generated by TextToCypherRetriever is invalid? Should we have a try/catch for the Cypher error?

Reply:
I've added a try except block which logs and re-raises the error in the event the LLM generates incorrect Cypher. Let me know if you think we need anything else for this scenario.

Reply:
We should think about the user experience here if the Cypher query isn't valid. Should we try generating a few more Cypher queries before throwing an error, at the expense of LLM calls? Should we throw an error immediately and advise the user to provide a more descriptive prompt or schema? I think we can try to make this nicer for the user

Reply:
At the moment, we could let the user decides what to do next IMHO. When we'll have agents or more complex retrieval strategies, we will be able to  deal with these failures.

Reply:
@willtai In the event Cypher query generation fails the user would need to either adjust the parameters of the LLM (temperature, etc.) or add in examples for the LLM to learn from. I can update the error message to be a bit more descriptive but I'm not sure that it'd be worth re-trying automatically as if query generation is failing, then there's probably something wrong with the prompt or the way the LLM is being used and the user will need to intervene.

Additionally, the idea currently is to not allow the users to edit the full prompt but to just to add examples the LLM can learn from. @oskarhane should we change this and allow users to edit the full prompt?

Reply:
I think I agree with @stellasia on this. 
For now we should let the developer know it failed and give them the full error message. Not sure what the pythonic way of doing this is, if it's to raise an exception and let the developer catch it or return something with a status field in it.

Reply:
I've done a bit of research and based on the advice in [here](https://eli.thegreenplace.net/2008/08/21/robust-exception-handling/) we should prefer raising exceptions to returning status codes. Currently I'm just logging the exception then re-raising it as a `RuntimeError`. We might want to think of specifying our own Exceptions at some point though as I imagine an LLM failing to return the correct response will be a common occurrence when using this library.

Reply:
I like that. I think we can improve our error handling using custom exceptions and that allows this application to be more scalable and maintainable to be used in other Python applications. I don't think that needs to block this PR though